src/roinput.rs

```diff
@@ -0,0 +1,856 @@
+    // The domain string length must be <= 20 (TODO: Limit at compile time)
+    fn domain_string(self, network_id: NetworkId) -> &'static str;
+}
+
+pub struct ROInput {
+    pub fields: Vec<PallasField>,
+    pub bits: BitVec<Lsb0, u8>,
+}
+
+impl ROInput {
+    pub fn new() -> Self {
+        ROInput {
+            fields: vec![],
+            bits: BitVec::new(),
+        }
+    }
+
+    pub fn append_field(&mut self, f: PallasField) {
+        self.fields.push(f);
+    }
+
+    pub fn append_scalar(&mut self, s: PallasScalar) {
+        // mina scalars are 255 bytes
+        let bytes = s.to_bytes(); // TODO: Combine these two into one-liner
+        let bits = &bytes.as_bits::<Lsb0>()[..PallasScalar::size_in_bits()];
+        self.bits.extend(bits);
+    }
+
+    pub fn append_bit(&mut self, b: bool) {
+        self.bits.push(b);
+    }
+
+    pub fn append_bytes(&mut self, bytes: Vec<u8>) {
+        self.bits.extend_from_bitslice(bytes.as_bits::<Lsb0>());
+    }
+
+    pub fn append_u32(&mut self, x: u32) {
+        self.append_bytes(x.to_le_bytes().to_vec());
+    }
+
+    pub fn append_u64(&mut self, x: u64) {
+        self.append_bytes(x.to_le_bytes().to_vec());
+    }
+
+    pub fn to_bytes(&mut self) -> Vec<u8> {
+        let mut bits: BitVec<Lsb0, u8> = self.fields.iter().fold(BitVec::new(), |mut acc, fe| {
+            acc.extend_from_bitslice(
+                &fe.to_bytes().as_bits::<Lsb0>()[..PallasField::size_in_bits()],
+            );
+
+            acc
+        });
+
+        bits.extend(self.bits.iter());
+
+        return bits.as_raw_slice().to_vec();
+    }
+
+    pub fn to_fields(&mut self) -> Vec<PallasField> {
+        let mut fields: Vec<PallasField> = self.fields.clone();
+
+        let bits_as_fields = self
+            .bits
+            .chunks(PallasField::size_in_bits() - 1)
+            .into_iter()
+            .fold(vec![], |mut acc, chunk| {
+                // Clone chunk into a clean copy so that the subsequent
+                // as_raw_slice() only contains the chunk's bits.
+                //
+                // A call to chunk.as_raw_slice() returns the bitvec's
+                // underlying memory as bytes, rather than a clean copy
+                // with only the chunk's bits set.  If chunk's size is
+                // not a byte-multiple, then the raw slice contains bits
+                // from the adjacent chunk.
+                //
+                // To get around this we explicitly allocate a new bitvec of
+                // the appropriate size, zero it and then copy the chunk's bits
+                // into it using clone_from_bitslice().
+                //
+                // N.B. BitVec::from_bitslice() doesn't work because it
+                // appears to copy chunk's raw memory into the new bitvec.
+
+                let mut bv = BitVec::<Lsb0, u8>::new();
+                bv.resize(chunk.len(), false);
+                bv.clone_from_bitslice(chunk);
```

Comment:
Yeah, it has to be 255.

Reply:
The reason I had to do this workaround is because if we access a non-byte aligned chunk directly, bitvec actually goes outside the chunk to include other bits from outside the chunk and, thus, gives the wrong value.  IOW, if chunk's size is not a byte-multiple, then the raw slice contains bits from the adjacent chunk.

Re raw... in this case it's what bitvec calls their function for to get the bytes for a bitvec (i.e. part of their public api https://docs.rs/bitvec/0.22.3/bitvec/) I didn't find any other way to convert back from bits to bytes.  Any ideas?

Reply:
Also wrt example code for 256 above... there's no `resize` method on BitSlice.

Reply:
To illustrate the problem from your sample code...  The code can be fixed with
```rust
        for chunk in self.bits.chunks(PallasField::size_in_bits() - 1) {
            // I think this will be 255 though, not 256 which might be needed to convert to bytes?
            let mut bytes = chunk.clone().to_bitvec();
            bytes.resize(PallasField::size_in_bits(), false);
            fields.push(PallasField::from_bytes(bytes.into()));
        }
```

But this fails to test cases due to the aforementioned bug.

To resolve the issue, the following works

```rust
        for chunk in self.bits.chunks(PallasField::size_in_bits() - 1) {
            // I think this will be 255 though, not 256 which might be needed to convert to bytes?
            let mut bv = BitVec::<Lsb0, u8>::new();
            bv.resize(chunk.len(), false);
            bv.clone_from_bitslice(chunk);
            bv.resize(PallasField::size_in_bits(), false);
            fields.push(PallasField::from_bytes(bv.into()));
        }
```

Reply:
> The reason I had to do this workaround is because if we access a non-byte aligned chunk directly, bitvec actually goes outside the chunk to include other bits that are not in the chunk and, thus, gives the wrong value. IOW, if chunk's size is not a byte-multiple, then the raw slice contains bits from the adjacent chunk.

yeah this is a problem with raw, you usually don't want to use them unless you're doing some pointer arithmetic or something

Reply:
Thanks, but even when I don't use raw it happens, so seems to be related to something else.  For example, it still happens with this code from above that has no raw anything...
```rust
        for chunk in self.bits.chunks(PallasField::size_in_bits() - 1) {
            // I think this will be 255 though, not 256 which might be needed to convert to bytes?
            let mut bytes = chunk.clone().to_bitvec();
            bytes.resize(PallasField::size_in_bits(), false);
            fields.push(PallasField::from_bytes(bytes.into()));
        }
```
Replacing the `clone()` with `clone_from_bitslice()` as in the second example I posted above seems to fix it.  Anyway, I'll switch to that version.

Reply:
doesn't that sound like a bug that should be reported upstream?